lib/xapi-stdext-encodings/encodings.ml

```diff
@@ -91,37 +69,31 @@ end
 module UTF8_CODEC (UCS_validator : UCS_VALIDATOR) = struct
   (* === Decoding === *)
 
-  let decode_header_byte byte =
-    if byte land 0b10000000 = 0b00000000 then (byte               , 1) else
-    if byte land 0b11100000 = 0b11000000 then (byte land 0b0011111, 2) else
-    if byte land 0b11110000 = 0b11100000 then (byte land 0b0001111, 3) else
-    if byte land 0b11111000 = 0b11110000 then (byte land 0b0000111, 4) else
-      raise UTF8_header_byte_invalid
-
   let decode_continuation_byte byte =
     if byte land 0b11000000 = 0b10000000 then byte land 0b00111111 else
       raise UTF8_continuation_byte_invalid
 
-  let width_required_for_ucs_value value =
-    if value < 0x000080 (* 1 lsl  7 *) then 1 else
-    if value < 0x000800 (* 1 lsl 11 *) then 2 else
-    if value < 0x010000 (* 1 lsl 16 *) then 3 else 4
-
-  let decode_character string index =
-    let value, width = decode_header_byte (Char.code string.[index]) in
-    let value = if width = 1 then (Int.of_int value)
-      else begin
-        let value = ref (Int.of_int value) in
-        for index = index + 1 to index + width - 1 do
-          let chunk = decode_continuation_byte (Char.code string.[index]) in
-          value := (!value <<< 6) ||| (Int.of_int chunk)
-        done;
-        if width > (width_required_for_ucs_value !value)
-        then raise UTF8_encoding_not_canonical;
-        !value
-      end in
-    UCS_validator.validate value;
-    (value, width)
+  let rec decode_continuation_bytes string last value index =
+    if index <= last then
+      let chunk = decode_continuation_byte (Char.code string.[index]) in
+      let value = (value lsl 6) lor chunk in
+      decode_continuation_bytes string last value (index + 1)
+    else value
+
+  let validate_character string index =
+    let value, width =
+      let byte = Char.code string.[index] in
+      if byte land 0b10000000 = 0b00000000 then (byte               , 1) else
+      if byte land 0b11100000 = 0b11000000 then (byte land 0b0011111, 2) else
```

Comment:
I know this is in existing code.. why are 2 bits flipped between these 2 cases, but only one in the successive ones?

Reply:
https://en.wikipedia.org/wiki/UTF-8 - the encoding table is helpful.